CRM/MembershipExtras/Service/PaymentPlanNextContributionDate.php

```diff
@@ -34,35 +39,69 @@ public function calculateAndUpdate() {
   }
 
   private function calculateNextScheduledContributionDate() {
-    $lastContributionReceiveDate = $this->getLastContributionReceiveDate();
-    if (empty($lastContributionReceiveDate)) {
+    $lastContribution = $this->getLastContribution();
+    if (empty($lastContribution)) {
       return NULL;
     }
 
-    $contributionRecur = civicrm_api3('ContributionRecur', 'get', [
-      'sequential' => 1,
-      'return' => ['id', 'start_date', 'frequency_interval', 'frequency_unit', 'cycle_day'],
-      'id' => $this->recurContributionID,
-      'options' => ['limit' => 1],
-    ])['values'][0];
-    $receiveDateCalculator = new CRM_MembershipExtras_Service_InstalmentReceiveDateCalculator($contributionRecur);
-    $receiveDateCalculator->setStartDate($lastContributionReceiveDate);
+    $relatedMemberships = $this->getRelatedMemberships($lastContribution['id']);
+    $atLeastOneFixedMembershipExist = FALSE;
+    foreach ($relatedMemberships as $membership) {
+      if ($membership['period_type'] == 'fixed') {
+        $atLeastOneFixedMembershipExist = TRUE;
+        $calculationMembership = $membership;
+        break;
+      }
+    }
+
+    if ($atLeastOneFixedMembershipExist) {
+      return date('Y-m-d', strtotime($calculationMembership['end_date'] . '00:00:00 +1 day'));
+    }
+    else {
+      $contributionRecur = civicrm_api3('ContributionRecur', 'get', [
+        'sequential' => 1,
+        'return' => ['id', 'start_date', 'frequency_interval', 'frequency_unit', 'cycle_day'],
+        'id' => $this->recurContributionID,
+        'options' => ['limit' => 1],
+      ])['values'][0];
 
-    return $receiveDateCalculator->calculate(2);
+      $receiveDateCalculator = new CRM_MembershipExtras_Service_InstalmentReceiveDateCalculator($contributionRecur);
+      $receiveDateCalculator->setStartDate($lastContribution['receive_date']);
+      return $receiveDateCalculator->calculate(2);
+    }
   }
 
-  private function getLastContributionReceiveDate() {
+  private function getLastContribution() {
     $lastContributionReceiveDate = civicrm_api3('Contribution', 'get', [
```

Comment:
Can you rename this variable to `lastContribution` like what you did in other places ?